tls/s2n_async_pkey.c

```diff
@@ -0,0 +1,387 @@
+
+    /* async_pkey will own the op, clean the mem blob to avoid freeing op */
+    GUARD_AS_RESULT(s2n_blob_init(&mem, NULL, 0));
+
+    ENSURE(conn->config->async_pkey_cb(conn, op) == S2N_SUCCESS, S2N_ERR_ASYNC_CALLBACK_FAILED);
+
+    /* Set state to waiting to allow op to be consumed by connection */
+    conn->handshake.async_state = S2N_ASYNC_INVOKED_WAITING;
+
+    /* Return an async blocked error to drop out of s2n_negotiate loop */
+    BAIL(S2N_ERR_ASYNC_BLOCKED);
+}
+
+S2N_RESULT s2n_async_pkey_decrypt_sync(struct s2n_connection *conn, struct s2n_blob *encrypted,
+                                       struct s2n_blob *init_decrypted, s2n_async_pkey_decrypt_complete on_complete)
+{
+    ENSURE_NONNULL(conn);
+    ENSURE_NONNULL(encrypted);
+    ENSURE_NONNULL(init_decrypted);
+    ENSURE_NONNULL(on_complete);
+
+    const struct s2n_pkey *pkey = conn->handshake_params.our_chain_and_key->private_key;
+
+    bool rsa_failed = s2n_pkey_decrypt(pkey, encrypted, init_decrypted) != S2N_SUCCESS;
+    GUARD_AS_RESULT(on_complete(conn, rsa_failed, init_decrypted));
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_async_pkey_sign(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                               struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
+{
+    ENSURE_NONNULL(conn);
+    ENSURE_NONNULL(digest);
+    ENSURE_NONNULL(on_complete);
+
+    if (conn->config->async_pkey_cb) {
+        GUARD_RESULT(s2n_async_pkey_sign_async(conn, sig_alg, digest, on_complete));
+    } else {
+        GUARD_RESULT(s2n_async_pkey_sign_sync(conn, sig_alg, digest, on_complete));
+    }
+
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_async_pkey_sign_async(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                                     struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
+{
+    ENSURE_NONNULL(conn);
+    ENSURE_NONNULL(digest);
+    ENSURE_NONNULL(on_complete);
+    ENSURE(conn->handshake.async_state == S2N_ASYNC_NOT_INVOKED, S2N_ERR_ASYNC_MORE_THAN_ONE);
+
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    GUARD_AS_RESULT(s2n_alloc(&mem, sizeof(struct s2n_async_pkey_op)));
+    GUARD_AS_RESULT(s2n_blob_zero(&mem));
+    struct s2n_async_pkey_op *op = ( void * )mem.data;
+
+    op->type = S2N_ASYNC_SIGN;
+    op->conn = conn;
+
+    struct s2n_async_pkey_sign_data *sign = &op->op.sign;
+    sign->on_complete                     = on_complete;
+    sign->sig_alg                         = sig_alg;
+
+    GUARD_AS_RESULT(s2n_hash_new(&sign->digest));
+    GUARD_AS_RESULT(s2n_hash_copy(&sign->digest, digest));
+
+    /* Block the handshake and set async state to invoking to block async states */
+    GUARD_AS_RESULT(s2n_conn_set_handshake_read_block(conn));
+    conn->handshake.async_state = S2N_ASYNC_INVOKING_CALLBACK;
+
+    /* async_pkey will own the op, clean the mem blob to avoid freeing op */
+    GUARD_AS_RESULT(s2n_blob_init(&mem, NULL, 0));
```

Comment:
`blob_init` allows to "forget" about an allocation? and prevent it from being released? This is ... interesting

Reply:
Yeah, essentially s2n_allocate allocates memory for s2n_blob, and DEFER_CLEANUP will execute s2n_free on the blob, when we exit the function. If you run s2n_blob_init on it it'll reset blob data and size and s2n_free will be a noop.

There is actually a memleak, as we allocate more memory in s2n_async_pkey_op for encrypted/decrypted/digest, so DEFER_CLEANUP actually needs to free op as op. Will send an update for that.